sdk/src/main/java/software/amazon/awssdk/iot/AwsIotMqttConnectionBuilder.java

```diff
@@ -500,6 +505,74 @@ public AwsIotMqttConnectionBuilder withWebsocketCredentialsProvider(CredentialsP
         return this;
     }
 
+    /**
+     * Configures the MQTT connection so it can use a custom authorizer.
+     * This function will modify the username, port, and TLS options.
+     *
+     * Note: All arguments are optional and can have "null" as valid input.
+     * See the description for each argument for information on what happens if null is passed.
+     * @param username The username to use with the custom authorizer. If null is passed, it will check to
+     *                 see if a username has already been set (via withUsername function). If no username is set then
+     *                 no username will be passed with the MQTT connection.
+     * @param authorizerName The name of the custom authorizer. If null is passed, then 'x-amz-customauthorizer-name'
+     *                       will not be added with the MQTT connection.
+     * @param signature The signature of the custom authorizer. If null is passed, then 'x-amz-customauthorizer-signature'
+     *                  will not be added with the MQTT connection.
+     * @param password The password to use with the custom authorizer. If null is passed, then no password will be set.
+     * @return
+     */
+    public AwsIotMqttConnectionBuilder withCustomAuthorizer(String username, String authorizerName, String signature, String password) {
+        isUsingCustomAuthorizer = true;
+        String username_string = "";
+        Boolean username_added_query = false;
+
+        if (username == null) {
+            if (config.getUsername() != null) {
+                username_string += config.getUsername();
+            }
+        } else {
+            username_string += username;
+        }
+
+        if (authorizerName != null) {
+            if (username_added_query == false) {
+                username_string += "?";
+                username_added_query = true;
+            } else {
+                username_string += "&";
+            }
+            if (authorizerName.contains("x-amz-customauthorizer-name=")) {
+                username_string += authorizerName;
+            } else {
+                username_string += "x-amz-customauthorizer-name=" + authorizerName;
+            }
+        }
+
+        if (signature != null) {
+            if (username_added_query == false) {
```

Comment:
This if-else matches exactly a block above.  I bet there's a way to refactor into a helper function that would make that go away.

In fact the larger if-statement may be refactorable: 
```add_username_parameter(..., "x-amz-customauthorizer-name", authorizerName)```
&
```add_username_parameter(..., "x-amz-customauthorizer-signature", signature)```
etc...